rfcs/20230418-constant-tensor-cache-capacity-control/README.md

```diff
@@ -0,0 +1,416 @@
+
+- In memory limited scenarios, we usually want to control the total memory
+  footprint, instead of the memory footprint of specific backend.
+- By the design of oneDNN Graph API, the API users don't need to be aware about
+  which backend is using. We don't like to break this design currently.
+- Backend selection is dynamic in mixed backend scenarios. Users can't
+  explicitly know which backend will run and how many partition will a backend
+  claims until graph partition stage done. So, it's hard to decide how many
+  memory a backend should use.
+
+### Capacity setting/getting frontend API
+Based on above considerations, we proposed to add the following two C++ API and
+corresponding C API to control the capacity of constant tensor cache:
+
+```cpp
+void set_constant_tensor_cache_capacity(dnnl::engine::kind eng_kind, size_t capacity);
+int get_constant_tensor_cache_capacity(dnnl::engine::kind eng_kind);
+```
+
+The unit of capacity is `MBytes`. The default capacity is
+`std::numeric_limits<size_t>::max()`, which means backends can cache constant
+tensors with no capacity limitation. Users can call the setter API to set the
+capacity for specific engine kind. If the cached tensors' size reaches the
+capacity, the new coming tensors won't be cached while the cached tensor won't
+be evicted by default. The getter API can be used to query the current capacity
+of specific engine kind. 
+
+We also allow to control the capacity by using an environment variable named
+`ONEDNN_CONSTANT_TENSOR_CACHE_CAPACITY`. The form of this env var value is
+`engine_kind:size`. We can use `;` to split the configuration for different
+engine kind. Below is an example:
+
+```shell
+# set cpu/gpu constant tensor cache capacity to 10 GB and 2 GB separately
+export ONEDNN_CONSTANT_TENSOR_CACHE_CAPACITY=cpu:10240;gpu:2048 
+```
+
+These two functional APIs could be called multiple times at runtime. But users
+should be aware that re-setting a smaller capacity will lead to eviction and
+impact performance. Setting capacity to 0 means to clear all cached tensors and
+disable constant tensor cache feature.
+
+The environment variable API should be set only once before the application
+start, then library will read the variable and cache it inside to reduce string
+parsing overhead. Re-setting environment variable at runtime will not take
+effect. Functional APIs have higher priority than environment variable API. If
+users call the functional APIs, it will overwrite the capacity value.
+
+With the new capacity control APIs, we can disable a constant tensor cache by
+setting its capacity to zero, and enable it by setting its capacity to non-zero.
+Users are suggested to use the new capacity control APIs. But for backward
+compatibility, the old enabling/disabling APIs will still be kept for a while.
+Calling the enabling/disabling APIs is equivalent to setting the cache capacity
+to default/zero value for all engine kinds.
```

Comment:
Since both APIs can control the enabling an disabling of constant tensor cache for now, we need to clarify the precedence of them.

Reply:
In my mind, the priority of new and old API are same. The old API will be re-implemented by calling new API inside.